app/ModelFunctions/LDAPFunctions.php

```diff
@@ -0,0 +1,377 @@
+<?php
+
+namespace App\ModelFunctions;
+
+use App\Exceptions\LDAPException;
+use App\Models\Configs;
+use App\Models\Logs;
+
+class LDAPFunctions
+{
+	/* @var resource $con holds the LDAP connection */
+	protected $con = null;
+
+	/* @var int $bound What type of connection does already exist? */
+	protected $bound = 0; // 0: anonymous, 1: user, 2: superuser
+
+	/* @var array user info queried from LDAP */
+	protected $user_info = [];
+
+	/**
+	 * Log debug message.
+	 *
+	 * @param string $method
+	 * @param string $line
+	 * @param string $text
+	 */
+	protected function _debug(string $method, string $line, string $text)
+	{
+		if ($this->getConf('debug', '0')) {
+			Logs::notice($method, $line, $text);
+		}
+	}
+
+	/**
+	 * Wraps around ldap_search, ldap_list or ldap_read depending on $scope.
+	 *
+	 * @param resource   $link_identifier
+	 * @param string     $base_dn
+	 * @param string     $filter
+	 * @param string     $scope           can be 'base', 'one' or 'sub'
+	 * @param array|null $attributes
+	 * @param int        $attrsonly
+	 * @param int        $sizelimit
+	 *
+	 * @return resource
+	 */
+	protected function _ldapsearch($link_identifier, $base_dn, $filter, $scope = 'sub', $attributes = null,
```

Comment:
Please add proper type hints to the parameters and also to the return type (if possible). E.g.

```suggestion
	protected function _ldapsearch($link_identifier, string $base_dn, $filter, string $scope = 'sub', array $attributes = [],
```

Again, this is only an example.